gson/src/main/java/com/google/gson/internal/bind/RecordTypeAdapterFactory.java

```diff
@@ -0,0 +1,480 @@
+        catch (NoSuchMethodException e) {
+            // We hit this either because we do not find the expected constructor, or we do not find a method
+            // name that matches the field name. In either of these classes, this is not a record, and we
+            // do not support serialization of it. The contract for TypeAdapterFactory is to return null for
+            // unsupported types. Gson will then attempt other means of creating a TypeAdapter. If all fails, an
+            // exception will be created where Gson was invoked to serialize/deserialize a type.
+            return null;
+        }
+    }
+
+    /**
+     * Factory to construct our own {@link RecordField} instances. This requires some reflection into the internal
+     * of Gson, which has logic to handle the {@link JsonAdapter} annotation. We also need to reflect into Gson
+     * to fetch the {@link Excluder}, that handles the {@link com.google.gson.annotations.Expose},
+     * {@link com.google.gson.annotations.Since} and {@link com.google.gson.annotations.Until} annotations. This
+     * allows us to closely follow the Gson mechanisms wrt. serialization/deserialization.
+     */
+    private static final class RecordFieldFactory {
+
+
+        private final Gson gson;
+        private final Excluder excluder;
+        private final JsonAdapterAnnotationTypeAdapterFactory jsonAdapterFactory;
+        private final ConstructorConstructor constructorConstructor;
+
+        public RecordFieldFactory(Gson gson, Excluder excluder,
+                JsonAdapterAnnotationTypeAdapterFactory jsonAdapterFactory,
+                ConstructorConstructor constructorConstructor) {
+            this.gson = gson;
+            this.excluder = excluder;
+            this.jsonAdapterFactory = jsonAdapterFactory;
+            this.constructorConstructor = constructorConstructor;
+        }
+
+        private RecordField createRecordField(Field field, Object recordComponent) {
+            TypeToken<?> fieldType = TypeToken.get(RECORD_HELPER.getGenericType(recordComponent));
+            JsonAdapter jsonAdapter = RECORD_HELPER.getAnnotation(recordComponent, JsonAdapter.class);
+            TypeAdapter<?> typeAdapter;
+            if (jsonAdapter == null) {
+                typeAdapter = gson.getAdapter(fieldType);
+            }
+            else {
+                typeAdapter = jsonAdapterFactory.getTypeAdapter(
+                        constructorConstructor,
+                        gson,
+                        fieldType,
+                        jsonAdapter
+                );
+            }
+
+            // Determine the serialized and deserialized names. If there is an SerializedName annotation
+            // present, we need to respect it. This might also mean that multiple keys acts as aliases
+            // when reading Json.
+            SerializedName serializedNameAnnotation =
+                    RECORD_HELPER.getAnnotation(recordComponent, SerializedName.class);
```

Comment:
Not sure if that actually works. `SerializedName` does not have `RECORD_COMPONENT` as target (and cannot have it as target if we want to still compile against Java 7).

However, you can obtain them from the private fields because they are propagated, see [Java Language Specification](https://docs.oracle.com/javase/specs/jls/se17/html/jls-8.html#jls-8.10.3):
> A component field is annotated with the annotations, if any, that appear on the corresponding record component and whose annotation interfaces are applicable in the field declaration context

Reply:
Oh, that's unfortunate. It's still worth adding support for records, but I suppose we won't be able to serialize their members using a name other than their own. One possible workaround would be to look for an annotation on the property method:

```
record Foo(int bar) {
  @SerializedName("baz")
  int bar() {
    return bar;
  }
}
```

Another possibility would be to remove the `@Target` restriction on `@SerializedName`, but that would be a bit sad.

I'm not sure I understand the comment about obtaining the annotation from the private field. If we could get `@SerializedName` onto the record component then it would indeed be copied onto the field, but how would we do that?

Anyway, I'm fine with this initial version not supporting names other than the Java name of the record component.

This is all making me think that we should have tests that run on Java ≥16, presumably using Maven profiles. Again that doesn't have to be in this PR, though perhaps you could add a test and then just exclude it for now. The exclusion could look like [this](https://www.codejava.net/tools/maven/maven-how-to-exclude-tests-from-build) and go [here](https://github.com/google/gson/blob/2154e468b490c8a6576db13eca31c9d1afe6bf78/gson/pom.xml#L85).

Reply:
> If we could get `@SerializedName` onto the record component then it would indeed be copied onto the field, but how would we do that?

Java allows to place `@SerializedName` on the record component because it has `FIELD` as target. It won't be available on the component then, but on the backing field.
Just try it out (e.g. with `jshell`):
```java
import java.lang.annotation.*;

@Target(ElementType.FIELD)
@Retention(RetentionPolicy.RUNTIME)
@interface MyAnnotation {}

R.class.getRecordComponents()[0].getAnnotations() // empty array
R.class.getDeclaredFields()[0].getAnnotations() // Annotation[1] { @MyAnnotation() }
```

The annotation is also propagated to the accessor method if it has `METHOD` as target, however only if it is _implicit_. If the user explicitly defines the accessor method the annotation won't be propagated. So the more reliable approach would be to obtain the annotation from the field.

Reply:
Oh, well that's OK. We should do that.

I've now configured the GitHub actions so that builds are done on both JDK 11 and JDK 17. So if we add a reflection-based test it should actually run on JDK 17. Ideally we would instead add a test that has actual records in it, and arrange for that to be excluded on Java 11. That would allow us to test the `@SerializedName` logic.

Reply:
Later if desired some of the [tests from my `gson-record-type-adapter-factory`](https://github.com/Marcono1234/gson-record-type-adapter-factory/blob/main/src/test/java/marcono1234/gson/recordadapter/SerializedNameTest.java) can be used. Though not all test cases are relevant for this implementation here.